src/nifs/mod.rs

```diff
@@ -9,222 +9,52 @@
-        let cross_terms: Vec<Vec<C::ScalarExt>> = (1..degree)
-            .map(|k| {
-                normalized.coeff_of(
-                    ColumnIndex::Challenge {
-                        column_index: r_index,
-                    },
-                    k,
-                )
-            })
-            .map(|multipoly| {
-                (0..num_row)
-                    .into_par_iter()
-                    .map(|row| data.eval(&multipoly, row))
-                    .collect::<Result<Vec<C::ScalarExt>, EvalError>>()
-            })
-            .collect::<Result<Vec<Vec<C::ScalarExt>>, EvalError>>()?;
-        let cross_term_commits: Vec<C> = cross_terms.iter().map(|v| ck.commit(v)).collect();
-        Ok((cross_terms, cross_term_commits))
-    }
-
-    /// Absorb all fields into RandomOracle `RO` & generate challenge based on that
-    pub(crate) fn generate_challenge(
-        pp_digest: C,
-        ro_acc: &mut RO,
-        U1: &RelaxedPlonkInstance<C>,
-        U2: &PlonkInstance<C>,
-        cross_term_commits: &CrossTermCommits<C>,
-    ) -> Result<<C as CurveAffine>::ScalarExt, Error> {
-        ro_acc.absorb_point(&pp_digest);
-        U1.absorb_into(ro_acc);
-        U2.absorb_into(ro_acc);
-        cross_term_commits
-            .iter()
-            .for_each(|cm| ro_acc.absorb_point(cm));
-
-        Ok(ro_acc.squeeze(NUM_CHALLENGE_BITS))
-    }
-
-    /// Generates a proof of correct folding using the NIFS protocol.
-    ///
-    /// This method takes two relaxed Plonk instance-witness pairs and calculates the folded instance and witness.
-    /// It also computes the cross terms and their commitments.
-    ///
-    /// # Arguments
-    /// * `ck`: The commitment key.
-    /// * `ro`: A mutable reference to the random oracle.
-    /// * `td`: Table data associated with the (strict not relaxed) plonk instance to be folded
-    /// * `U1`: The first relaxed Plonk instance.
-    /// * `W1`: The witness for the first relaxed Plonk instance.
-    ///
-    /// # Returns
-    /// A tuple containing the NIFS instance and the folded relaxed Plonk instance-witness pair.
-    ///
-    /// # Context
-    /// The prove function is central to the NIFS protocol. It demonstrates
-    /// that two polynomial relations have been correctly folded into one,
-    /// providing cryptographic evidence of this fact.
-    pub fn prove(
+pub mod vanilla;
+
+/// Trait representing the NIFS folding scheme.
+trait FoldingScheme<C: CurveAffine, RO: ROTrait<C>> {
+    // metadata for prover including pp_digest
+    type ProverParam;
+    // metadata for verifier including pp_digest
+    type VerifierParam;
+    // Accumulator contains AccumulatorInstance and Witness data
+    type Accumulator;
+    type AccumulatorInstance;
+    fn setup_params(
+        td: &TableData<C::ScalarExt>,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), Error>;
+    // Perform the folding operation as a prover.
+    fn prove(
```

Comment:
Please add empty lines between different trait types and methods, it's hard to read them in such a heap